test/rules.jl

```diff
@@ -31,8 +31,42 @@ _second(t) = Base.tuple_type_head(Base.tuple_type_tail(t))
 
     frx, fr = frule(cool, 1)
     @test frx == 2
-    @test fr(1) == 1
-    rrx, rr = rrule(cool, 1)
+    @test fr(NamedTuple(), 1) == (1,)
+    rrx, (rr) = rrule(cool, 1)
+    self, rr1 = rr(1)
+    @test self == NO_FIELDS
     @test rrx == 2
-    @test rr(1) == 1
+    @test rr1 == 1
+end
+
+
+@testset "Wirtinger scalar_rule" begin
+    myabs2(x) = abs2(x)
+    @scalar_rule(myabs2(x), Wirtinger(x', x))
+
+    # real input
+    x = rand(Float64)
+    f, pushforward = frule(myabs2, x)
+    @test f === x^2
+
+    df = @inferred pushforward(NamedTuple(), One())
```

Comment:
minor, but I was confused by this test, 
becuase at first I thought `pushforward` was a generic `ChainRulesCore` function and couldn't see how it was getting info about `f = myabs2`... can we rename it `f_pushforward` or `myabs2_pushforward` or somesuch?

I know it's only test code, but I often read tests for understanding how the package is meant to be used :)

Reply:
by the way, naming gripe aside, this testset is a great little tutorial!